src/System.Collections/src/System/Collections/Generic/EqualityComparer.cs

```diff
@@ -0,0 +1,336 @@
+
+                    if (typeof(T) == typeof(SByte))
+                        comparer = new EqualityComparerForSByte();
+                    else if (typeof(T) == typeof(Byte))
+                        comparer = new EqualityComparerForByte();
+                    else if (typeof(T) == typeof(Int16))
+                        comparer = new EqualityComparerForInt16();
+                    else if (typeof(T) == typeof(UInt16))
+                        comparer = new EqualityComparerForUInt16();
+                    else if (typeof(T) == typeof(Int32))
+                        comparer = new EqualityComparerForInt32();
+                    else if (typeof(T) == typeof(UInt32))
+                        comparer = new EqualityComparerForUInt32();
+                    else if (typeof(T) == typeof(Int64))
+                        comparer = new EqualityComparerForInt64();
+                    else if (typeof(T) == typeof(UInt64))
+                        comparer = new EqualityComparerForUInt64();
+                    else if (typeof(T) == typeof(IntPtr))
+                        comparer = new EqualityComparerForIntPtr();
+                    else if (typeof(T) == typeof(UIntPtr))
+                        comparer = new EqualityComparerForUIntPtr();
+                    else if (typeof(T) == typeof(Single))
+                        comparer = new EqualityComparerForSingle();
+                    else if (typeof(T) == typeof(Double))
+                        comparer = new EqualityComparerForDouble();
+                    else if (typeof(T) == typeof(Decimal))
+                        comparer = new EqualityComparerForDecimal();
+                    else if (typeof(T) == typeof(String))
+                        comparer = new EqualityComparerForString();
+                    else
+                        comparer = new LastResortEqualityComparer<T>();
+
+                    _default = (EqualityComparer<T>)comparer;
+                }
+
+                return _default;
+            }
+        }
+
+        private static volatile EqualityComparer<T> _default;
+
+        public abstract bool Equals(T x, T y);
+
+        public abstract int GetHashCode(T obj);
+
+        int IEqualityComparer.GetHashCode(object obj)
+        {
+            if (obj == null)
+                return 0;
+            if (obj is T)
+                return GetHashCode((T)obj);
+            throw new ArgumentException(SR.Argument_InvalidArgumentForComparison);
+        }
+
+        bool IEqualityComparer.Equals(object x, object y)
+        {
+            if (x == y)
+                return true;
+            if (x == null || y == null)
+                return false;
+            if ((x is T) && (y is T))
+                return Equals((T)x, (T)y);
+            throw new ArgumentException(SR.Argument_InvalidArgumentForComparison);
+        }
+    }
+
+    //
+    // ProjectN compatiblity notes:
```

Comment:
Is this comment still relevant?  And this should be ".NET Native" rather than "ProjectN", right?